### PR TITLE
adding extra data fields to the bus location records

### PR DESCRIPTION
**Pull request by anthonymobile**
pulling headsign, etc. from the XML now since its there and useful for other applications like signs + timestamping the fetch time into each record seems to make sense for data integrity.

untested.

### CODE REVIEW

**Comment by anthonymobile** [comment]
ps. this is my 1st ever pull request. so not sure i do it right.

File: njbuses/Buses.py [hunk 1]
```diff
@@ -33,14 +37,30 @@ def parse_xml_data(data):
...
     if source not in _sources:
         raise AssertionError('Unknown source=%s valid sources=%s' % (key, str(sources.keys())))
 
+    # mark the time right before we fetch the source data
+    import datetime
+    mark = datetime.datetime.now()
+    
     import urllib2
     data = urllib2.urlopen(_sources[source]).read()
     return parse_xml_data(data)
```
Comment (reyale) [inline njbuses/Buses.py]: mark time should be passed to parse_xml_data as an optional parameter.

File: njbuses/Buses.py [hunk 2]
```diff
@@ -33,14 +37,30 @@ def parse_xml_data(data):
             continue
         lon = lons[0].text
 
-        results.append(Bus(id_text, lat, lon))
+    headsign = atype.findall('fs')
+    if len(fs) != 1:
```
Comment (reyale) [inline njbuses/Buses.py]: Probably should be: len(headsign)

File: njbuses/Buses.py [hunk 3]
```diff
@@ -33,14 +37,30 @@ def parse_xml_data(data):
...
+            continue
+        headsign = fs[0].text
+
+    destination = atype.findall('pd')
+    if len(pd) != 1:
```
Comment (reyale) [inline njbuses/Buses.py]: probably should be len(destination)

File: njbuses/Buses.py [hunk 4]
```diff
@@ -33,14 +37,30 @@ def parse_xml_data(data):
...
+            continue
+        destination = pd[0].text
+
+    timestamp = mark
+
+    results.append(Bus(id_text, lat, lon))
```
Comment (reyale) [inline njbuses/Buses.py]: Bus now requires your new parameters.

File: njbuses/Buses.py [hunk 5]
```diff
@@ -33,14 +37,30 @@ def parse_xml_data(data):
...
+    if len(fs) != 1:
+            continue
+        headsign = fs[0].text
+
+    destination = atype.findall('pd')
+    if len(pd) != 1:
+            continue
+        destination = pd[0].text
+
+    timestamp = mark
```
Comment (reyale) [inline njbuses/Buses.py]: Reassignment unneeded, rename as part of function parameter.

File: njbuses/Buses.py [hunk 6]
```diff
@@ -33,14 +37,30 @@ def parse_xml_data(data):
             continue
         lon = lons[0].text
 
-        results.append(Bus(id_text, lat, lon))
+    headsign = atype.findall('fs')
```
Comment (reyale) [inline njbuses/Buses.py]: These lines should be inside the for-loop.

**Comment by anthonymobile** [comment]
thanks. think i got most of those.

**Comment by anthonymobile** [comment]
next going to write a grabber to cron-ify fetching this 1x minute (for starters) into a db.

**Comment by reyale** [comment]
I've merged these changes in, fixed the code and tested it.  Two  things.  1) Made fields other than ID, LAT, LON optional (will save blank if they don't exist).  2) I make sure the timestamp is as close to getting the data as possible.

**Comment by anthonymobile** [comment]
having a hard time seeing what changed but still working up to speed on git. next onto the grabber. going to try to get it working over the weekend. trying to figure out what the right frequency for grabs is. 1 minute is good for service auditing but up to 1 second would be useful for seeing how drivers avoid obstacles / approach curbs+stops, etc. (assuming the GPS and reporting is that precise, consistent + timely)

**Comment by anthonymobile** [comment]
1) Made fields other than ID, LAT, LON optional (will save blank if they don't exist). 

> what changed for this?
> 2) I make sure the timestamp is as close to getting the data as possible.
> not sure how this changed either?

**Comment by reyale** [comment]
Maybe you're not seeing the changes because you didn't update your local copy from the master.  Here's the copy I am looking at: https://github.com/reyale/openjerseycity/blob/master/njbuses/Buses.py#L45

1) I removed the `continue` use from headsign and direction.  Previously you'd "continue" and skip adding them to the results array if that information wasn't available.
2) I moved the `datetime.datetime.now()`after all the imports.  Imports are kind of caustly in python because it lazy-loads the modules it uses.

**Comment by reyale** [comment]
About the grabber.  I think the main issue with this data is that we don't know where the bus stops are.  The only way I know that this data-set could help us here is if it's granular and accurate enough that it can detect when the bus lingers in one position for a length of time.  That'd be a candidate stop.  Unfortunately buses stop for all sorts of reasons, but we might be able to categorize them.  For this to work we need pretty granular data.  I suggest running the grabber with just lat+lon+id data every second for a few hours and tracing a bus or two to see if the lat+lon data is really updating on that granularity.   Hopefully they won't shut down this endpoint with our spamming.   

From a location accuracy perspective I don't think this GPS is THAT accurate.  I definitely see some buses go off the curb by a meter or two.

**Comment by anthonymobile** [comment]
i was running it every minute for 3 months from the same IP and they didn’t notice, so….

but yes, ill grab this sample and package it up.

if you think of anything else that makes sense…

**Pull request merged**